File: CppRefl.UnitTest.Cpp/Tests/EnumTests.cpp
#include "gtest/gtest.h"

#include "ReflectedCode/EnumCode.h"
#include "Reflection/EnumInfo.h"

TEST(EnumTests, EnumTypes)
{
	EXPECT_STREQ(cpprefl::GetReflectedType<ReflectedEnum>().mName, "ReflectedEnum");
	EXPECT_EQ(&cpprefl::ReflectedEnum<ReflectedEnum>().mType, &cpprefl::GetReflectedType<ReflectedEnum>());
}

TEST(EnumTests, EnumValues)
{
	const auto& enumInfo = cpprefl::ReflectedEnum<ReflectedEnum>();

	EXPECT_EQ(enumInfo.mValues.size(), 3);

	EXPECT_STREQ(enumInfo.mValues[0].mName, "EnumValue1");
	EXPECT_EQ(enumInfo.mValues[0].mValue, (int)ReflectedEnum::EnumValue1);

	EXPECT_STREQ(enumInfo.mValues[1].mName, "Bar");
	EXPECT_EQ(enumInfo.mValues[1].mValue, (int)ReflectedEnum::Bar);

	EXPECT_STREQ(enumInfo.mValues[2].mName, "Foo");
	EXPECT_EQ(enumInfo.mValues[2].mValue, (int)ReflectedEnum::Foo);
}

File: Compiler/CppRefl.Compiler.Tests/EnumTests.cs
using CppRefl.CodeGeneration.Reflection;

namespace CppRefl.Compiler.Tests
{
	internal class EnumTests : CompilerTestBase
	{
		protected override string ReflectionFile => Path.Combine(TestUtil.Const.Runtime.Tests.ReflectedHeadersDirectory.FullName, "EnumCode.h");

		[Test]
		public void EnumTypes()
		{
			TestUtil.TestType(Registry.GetType("ReflectedEnum")!, "ReflectedEnum", TypeKind.Enum, isEnum: true);
			TestUtil.TestType(Registry.GetType("TestNamespace::NamespacedEnum")!, "NamespacedEnum", TypeKind.Enum, isEnum: true, typeNamespace: "TestNamespace");

			Assert.That(Registry.GetEnum("ReflectedEnum")!.Type, Is.EqualTo(Registry.GetType("ReflectedEnum")));
			Assert.That(Registry.GetEnum("TestNamespace::NamespacedEnum")!.Type, Is.EqualTo(Registry.GetType("TestNamespace::NamespacedEnum")));
		}

		[Test]
		public void EnumValues()
		{
			{
				var enumInfo = Registry.GetEnum("ReflectedEnum")!;
				Assert.That(enumInfo.Values.Count, Is.EqualTo(3));

				Assert.That(enumInfo.Values[0].Name, Is.EqualTo("EnumValue1"));
				Assert.That(enumInfo.Values[0].Value, Is.EqualTo(0));

				Assert.That(enumInfo.Values[1].Name, Is.EqualTo("Bar"));
				Assert.That(enumInfo.Values[1].Value, Is.EqualTo(10));

				Assert.That(enumInfo.Values[2].Name, Is.EqualTo("Foo"));
				Assert.That(enumInfo.Values[2].Value, Is.EqualTo(11));
			}

			{
				var enumInfo = Registry.GetEnum("TestNamespace::NamespacedEnum")!;
				Assert.That(enumInfo.Values.Count, Is.EqualTo(2));

				Assert.That(enumInfo.Values[0].Name, Is.EqualTo("Valu1"));
				Assert.That(enumInfo.Values[0].Value, Is.EqualTo(0));

				Assert.That(enumInfo.Values[1].Name, Is.EqualTo("Valu2"));
				Assert.That(enumInfo.Values[1].Value, Is.EqualTo(1));
			}
		}
	}
}
